### ISSUE TITLE
I’m implementing pure rust client let’s merge our efforts 

### ISSUE THREAD

**Issue by asahaf** [influxdata/rskafka#122]
Great job!

I was frustrated using the C based lib with binding so I decided to write a pure rust tokio v1 based general purpose Kafka client. 
The client supports Kafka 0.11 and above. 

The client is still no complete yet but I’ve made some progress. 

Let’s contribute and merge our efforts.

**Comment by crepererum** [comment]
Hey @asahaf,

thanks for reaching out. We at InfluxData very much welcome contributions and -- where they make sense -- collaborations. For that I have a few questions:

- What is the state of your crate?
- What type of collaboration you are envisioning?
- What is your mid to long-term vision for resulting artifact?

We could also have quick Zoom call if that would help.

**Comment by asahaf** [comment]
Sure we can have a Zoom call. This would give me a chance to explain the design. Here is a brief of the design.

At the lowest level I have **broker connection** which runs as an independent Tokio task, run indefinitely in a loop until it shutdown. It automatically recovers from network dis-connectivity, does graceful shutdown, queuing requests and it handles all the low level networking with the broker (sending and receiving). The broker connection receives requests through a channel for requests to be sent to the broker and sends the received responses back through a one-shot channel created for each request. I'm using channels to avoid using any mutex and locking. The broker connection provides two interfaces to communicate with. One channel for sending requests (data channel), and one channel for sending commands (control channel, used to do graceful shutdown, pauses, etc). Those channels are used by the Kafka Client which I'll talk about in the next.

On top of the **broker connection** I have Kafka Client which maintains a list of those broker connections, I mentioned in the previous paragraph, one for each cluster node. Kafka client it acts as a proxy for those connections. It is used by the upper clients (will talk about later) to direct the requests to the recipient brokers through broker connections. This kafka client is consider as a mid-level client and it only should be used by the upper clients like Admin Client, Consuming Client, Producing Client, etc. Kafka Client deals with plain Kafka requests and responses.

User facing clients, like Admin Client, Consuming Client, and Producing Clients are built on top of the Kafka Client. Those upper clients provide APIs to perform different tasks, like creating a topic, deleting a topic, consuming a topic, etc. and internally those clients create the required Kafka API requests (as some of those actions required multiple Kafka requests), then send them through the Kafka Client which they are built on top. 

The state of the crate is, I already finished the broker connection, the Kafka Client, and finish a framework for encoding and decoding the requests and responses. Currently I'm working on the upper clients, starting with Consuming Client as it's more challenging than the other clients. I'm implementing currently the group management and the assignments. The idea is to make a minimum viable clients that I planning to keep adding the missing pieces until it's complete.

All of this is one man show, and I've spent a lot of hours and days reading Kafka documentations, KIPs and even the source code of their java client, trying to understanding and putting the design for this async client. Contributions definitely welcomed and needed.

I've gone through the source code of your client and I see you mentioned that you developed it with limited scope. I'm not sure if you're interested to have a complete Kafka client which does beyond what you've done already. 

My long-term vision of the resulting artifact is a complete Kafka client, a decent pure Rust Kafka client that helps the community. The contribution I'm looking for is maybe improving the design, implementing the remaining upper clients, do some security checks, and testing as it's very important. I'm not an expert in all areas, but trying my best.

If I got you interested, lets have a Zoom call :)
